google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.gson.Gson;
+import com.google.gson.JsonParseException;
+import com.google.gson.reflect.TypeToken;
+
+import com.intellij.icons.AllIcons.General;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.remoteServer.runtime.deployment.DeploymentRuntime;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.Type;
+import java.util.List;
+
+import javax.swing.SwingUtilities;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * Takes care of undeploy and stop of App Engine applications.
+ */
+class AppEngineDeploymentRuntimeImpl extends DeploymentRuntime {
+
+  private static final String STOP_CONFIRMATION_URI_OPEN_TAG =
+      "<a href='https://cloud.google.com/appengine/docs/java/console/#versions'>";
+  private static final String STOP_CONFIRMATION_URI_CLOSE_TAG = "</a>";
+
+  private Project project;
+  private AppEngineHelper appEngineHelper;
+  private LoggingHandler loggingHandler;
+  private DeployOutput deployOutput;
+
+  public AppEngineDeploymentRuntimeImpl(
+      @NotNull Project project,
+      @NotNull AppEngineHelper appEngineHelper,
+      @NotNull LoggingHandler loggingHandler,
+      @NotNull String deploymentOutput) {
+    this.project = project;
+    this.appEngineHelper = appEngineHelper;
+    this.loggingHandler = loggingHandler;
+
+    try {
+      deployOutput = parseDeployOutputToService(deploymentOutput);
+    } catch (JsonParseException e) {
+      loggingHandler.print(
+          GctBundle.message("appengine.deployment.version.extraction.failure") + "\n");
+      Logger.getInstance(AppEngineDeploymentRuntimeImpl.class).warn(
+          "Could not retrieve service/version info of deployed application", e);
+      deployOutput = null;
```

Comment:
I think etan's suggestion to do the parsing in the action and construct an
appropriate runtime based off of the result of that could work.

On Tue, May 3, 2016 at 1:18 PM, Chanseok Oh <notifications@github.com>
wrote:

> In
> google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/660#discussion_r61919422>
> :
>
> > +      @NotNull Project project,
> > +      @NotNull AppEngineHelper appEngineHelper,
> > +      @NotNull LoggingHandler loggingHandler,
> > +      @NotNull String deploymentOutput) {
> > +    this.project = project;
> > +    this.appEngineHelper = appEngineHelper;
> > +    this.loggingHandler = loggingHandler;
> > +
> > +    try {
> > +      deployOutput = parseDeployOutputToService(deploymentOutput);
> > +    } catch (JsonParseException e) {
> > +      loggingHandler.print(
> > +          GctBundle.message("appengine.deployment.version.extraction.failure") + "\n");
> > +      Logger.getInstance(AppEngineDeploymentRuntimeImpl.class).warn(
> > +          "Could not retrieve service/version info of deployed application", e);
> > +      deployOutput = null;
>
> I agree on that JSON parsing failure is still a successful deployment,
> because it is true that the app is actually deployment and running in this
> case.
>
> As @etanshaul <https://github.com/etanshaul> said, even if we move the
> parsing to DeployAction, we will still have to construct (possibly another)
> DeploymentRuntime anyway to override isUndeploySupported to return false.
> In this sense, I think it is not a bad idea to do parsing in this
> constructor. Ideas?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/660/files/5f2176fab1a2f91586c3eb81b376971583718b4a#r61919422>
>